modules/websocket/websocket-destination.c

```diff
@@ -0,0 +1,259 @@
+{
+  WebsocketDestDriver *self = (WebsocketDestDriver *)d;
+  return &self->template_options;
+}
+
+static void
+websocket_dd_free(LogPipe *destination)
+{
+  WebsocketDestDriver *self = (WebsocketDestDriver *) destination;
+  g_free(self->template);
+  g_free(self->address);
+  g_free(self->path);
+  g_free(self->protocol);
+  g_free(self->mode);
+  g_free(self->cert);
+  g_free(self->key);
+  g_free(self->cacert);
+  log_threaded_dest_driver_free(destination);
+}
+
+
+static void
+websocket_worker_thread_init(LogThrDestDriver *destination)
+{
+  WebsocketDestDriver *self = (WebsocketDestDriver *)destination;
+
+  msg_debug("Worker thread started",
+            evt_tag_str("driver", self->super.super.super.id),
+            NULL);
+
+  if (strcmp(self->mode, "client") == 0)
+    websocket_client_create(self->protocol, self->address, self->port,
+      self->path, self->client_use_ssl_flag, self->cert, self->key,
+      self->cacert, &self->msgqid, &self->service_pid);
+  else if (strcmp(self->mode, "server") == 0)
+    websocket_server_create(self->protocol, self->port,
+      self->client_use_ssl_flag, self->cert, self->key, self->cacert, NULL,
+      &self->msgqid, &self->service_pid);
+}
+
+
+static void
+websocket_dd_disconnect(LogThrDestDriver *destination)
+{
+  WebsocketDestDriver *self = (WebsocketDestDriver *)destination;
+  if (strcmp(self->mode, "client") == 0)
+    websocket_client_disconnect(self->service_pid);
+  else if (strcmp(self->mode, "server") == 0)
+    websocket_server_shutdown(self->service_pid);
+}
+
+static worker_insert_result_t
+websocket_worker_insert(LogThrDestDriver *destination, LogMessage *msg)
+{
+  WebsocketDestDriver *self = (WebsocketDestDriver *)destination;
+  GString *result = g_string_new("");
+  log_template_format(self->template, msg, &self->template_options, LTZ_LOCAL, self->super.seq_num, NULL, result);
+
+  int res;
+  if (strcmp(self->mode, "client") == 0)
+    res = websocket_client_send_msg(result->str, self->msgqid, self->port);
+  else if (strcmp(self->mode, "server") == 0)
+    res = websocket_server_broadcast_msg(result->str, self->msgqid, self->port);
+
+  g_string_free(result, TRUE);
+  if (res == 0)
+    return WORKER_INSERT_RESULT_SUCCESS;
+  else
+    return WORKER_INSERT_RESULT_ERROR;
+}
+
+static gchar *
+websocket_dd_format_stats_instance(LogThrDestDriver *d)
+{
+  static gchar persist_name[1024];
+  WebsocketDestDriver *self = (WebsocketDestDriver *)d;
+
+  g_snprintf(persist_name, sizeof(persist_name), "websocket_dest:%s:%s:%d:%s:%s",
+    self->mode, self->address, self->port, self->protocol, self->path);
+  return persist_name;
+}
+
+static gchar *
+websocket_dd_format_persist_name(LogThrDestDriver *d)
+{
+  static gchar persist_name[1024];
+  WebsocketDestDriver *self = (WebsocketDestDriver *)d;
+
+  g_snprintf(persist_name, sizeof(persist_name), "websocket_dest:%s:%s:%d:%s:%s",
+    self->mode, self->address, self->port, self->protocol, self->path);
+  return persist_name;
+}
```

Comment:
The two functions are combined into one.
The same problem in ZMQ is fixed now !